qemu/tests/single_driver_update.py

```diff
@@ -0,0 +1,170 @@
+        driver_repo_path = params.get("driver_repo_path")
+        driver_name = params.get("driver_name")
+        for folder in folders:
+            uninstall_command = driver_uninstall_command % (
+                driver_repo_path,
+                folder,
+                driver_name)
+            error.context("uninstall cli %s"
+                          % uninstall_command, logging.info)
+            status, output = session.cmd_status_output(uninstall_command)
+            if status == 1073741824:
+                logging.info("reboot is required")
+                session = vm.reboot(session)
+            elif status == 0:
+                logging.info("driver uninstalled successfully")
+            elif status == -2147483648:
+                logging.info("No Driver Store Entry %s"
+                             % driver_uninstall_command)
+                session = vm.reboot(session)
+            else:
+                get_installation_logs(session)
+                raise error.TestError("unknown error %s"
+                                      % status, logging.info)
+
+    def driver_install(session, folder, downgrade=False):
+        install_cmd = params.get("driver_install_command")
+        tmp_folder = params.get("tmp_folder", 'C:\\tmp\\')
+        winutils = utils_misc.get_winutils_vol(session)
+        if downgrade:
+            tmp_folder = winutils + ":"
+        install_cmd = install_cmd % (winutils, tmp_folder, folder)
+        error.context("install %s" % install_cmd, logging.info)
+        status, output = session.cmd_status_output(install_cmd)
+        if status == -2147483648:
+            msg = "Testing Failed"
+            msg += "it is driver not digital signed by either"
+            msg += "Microsoft or Redhat "
+            msg += "pls report a bug agaist virtio-win component"
+            get_installation_logs(session)
+            raise error.TestError("%s, %s" % (msg, output), logging.info)
+        elif status == 1:
+            logging.info("driver installed successfully %s"
+                         % output)
+            logging.debug("yonit bitmap benchmark was not found")
+        elif status == 0:
+            logging.info("driver downgrade successfully %s"
+                         % output)
+        else:
+            msg = "return %s ,pls check C:\windows\dpinst.log for help" % status
+            get_installation_logs(session)
+            raise error.TestError("%s, %s" % (msg, output), logging.info)
+
+    def get_current_version(session):
+        device_name = params.get("device_name")
+        driver_version_command = params.get("driver_version_command")
+        device_name_quota = "\"%s\"" % device_name
+        exec_dri_version = driver_version_command % device_name_quota
+        status, outputfull = session.cmd_status_output(exec_dri_version)
+        if status == 0:
+            key = "\d*\.\d*\.\d*\.\d*"
+            output = re.findall(key, outputfull, re.M)[-1]
```

Comment:
output = re.findall(key, outputfull, re.M)[-1]
  IndexError: list index out of range